go.mod

```diff
@@ -1,22 +1,41 @@
 module github.com/caddyserver/ingress
 
-go 1.12
+go 1.13
 
 require (
-	github.com/caddyserver/caddy/v2 v2.0.0-beta6
-	github.com/docker/distribution v2.7.1+incompatible // indirect
-	github.com/mholt/certmagic v0.7.5
-	github.com/opencontainers/go-digest v1.0.0-rc1 // indirect
+	github.com/caddyserver/caddy/v2 v2.0.0-beta11
+	github.com/mholt/certmagic v0.9.0
 	github.com/pkg/errors v0.8.1
 	github.com/prometheus/client_golang v1.1.0
 	github.com/sirupsen/logrus v1.4.2
 	gopkg.in/go-playground/assert.v1 v1.2.1 // indirect
 	gopkg.in/go-playground/pool.v3 v3.1.1
-	k8s.io/api v0.0.0-20190602125759-c1e9adbde704
-	k8s.io/apiextensions-apiserver v0.0.0-20190602131520-451a9c13a3c8 // indirect
-	k8s.io/apimachinery v0.0.0-20190602125621-c0632ccbde11
-	k8s.io/client-go v0.0.0-20190602130007-e65ca70987a6
-	k8s.io/cloud-provider v0.0.0-20190503112208-4f570a5e5694 // indirect
-	k8s.io/kubernetes v1.14.1
-	k8s.io/utils v0.0.0-20190506122338-8fab8cb257d5 // indirect
+	k8s.io/api v0.17.0
+	k8s.io/apimachinery v0.17.0
+	k8s.io/client-go v0.17.0
+	k8s.io/kubernetes v1.17.0
+)
+
+replace (
+	k8s.io/api => k8s.io/api v0.17.0
+	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.17.0
+	k8s.io/apimachinery => k8s.io/apimachinery v0.17.0
+	k8s.io/apiserver => k8s.io/apiserver v0.17.0
+	k8s.io/cli-runtime => k8s.io/cli-runtime v0.17.0
+	k8s.io/client-go => k8s.io/client-go v0.17.0
+	k8s.io/cloud-provider => k8s.io/cloud-provider v0.17.0
+	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.17.0
+	k8s.io/code-generator => k8s.io/code-generator v0.17.0
+	k8s.io/component-base => k8s.io/component-base v0.17.0
+	k8s.io/cri-api => k8s.io/cri-api v0.17.0
+	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.17.0
+	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.17.0
+	k8s.io/kube-controller-manager => k8s.io/kube-controller-manager v0.17.0
+	k8s.io/kube-proxy => k8s.io/kube-proxy v0.17.0
+	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.17.0
+	k8s.io/kubectl => k8s.io/kubectl v0.17.0
+	k8s.io/kubelet => k8s.io/kubelet v0.17.0
+	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.17.0
+	k8s.io/metrics => k8s.io/metrics v0.17.0
+	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.17.0
 )
```

Comment:
Can you help me understand why these are necessary?

Reply:
@mholt do you mena the `replace` section? Since K8s is moving from an outdated dependency system to go mod and they are using a single repo (that should have a single tag) but at the same time they chose to use a 0.x.y version of go mods (see https://github.com/kubernetes/kubernetes/issues/84608#issuecomment-564322329 ) we need to explain dependencies that are reported as `v0.0.0` in the `kubernetes/kubernetes` repo (since at the moment the reference are not updated for each release).

I don't have a background that allow me to go deep on explanation, you can find a lot of info on SIG-ARCHITECTURE https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/2019-03-19-go-modules.md

Other project follow the same approach (see -for example- https://github.com/kubernetes/ingress-nginx/blob/master/go.mod#L58-L81 )

Reply:
Huh... that is really unusual. So this is a temporary thing, it sounds like?

Reply:
@mholt I hope too, but sincerely I have no idea on schedule / improvement for next releases on this side...